### DISCUSSION TITLE
NMEA 2000 compass with IMU

### DISCUSSION THREAD

**Discussion by mgrouch** [bareboat-necessities/lysmarine_gen#447] [General]
https://github.com/htool/ESP32_Precision-9_compass_CMPS14

**Comment by mgrouch** [comment]
Another idea is to make one with m5stack AtomS3R

**Reply by tgsail1** [reply]
I am trying to do exactly this. I've got the sensor fusion working, but the original NMEA2000_esp32 code does not support GPIO pins 5 and 6 which are used in the S3 chips. I'm trying to use your version of the NMEA2000_esp32 code which is used in bbn-m5-s3-n2k-i2c.ino. I can't seem to get that arduino code to compile using the M5Atoms3 board definition. Any suggestions as to why the code might not compile? I've already made sure that the #include statements explicitly point to your versions of the library.

**Reply by mgrouch** [reply]
You can see exact versions and how it's built in GitHub action

https://github.com/bareboat-necessities/bbn-m5-s3-n2k-i2c/blob/main/.github/workflows/build.yaml

Making fusion work is not enough you need to make good calibration work with elliptical fit.

**Reply by mgrouch** [reply]
Working compass on atomS3R is here

https://github.com/bareboat-necessities/bbn_compass

**Reply by tgsail1** [reply]
The compass works well enough, what doesn't work is the NMEA2000 interface.

**Comment by tgsail1** [comment]
I was able to get this to work on arduino 2.3.7 IDE by downloading your GitHub files in a zip folder and loading it as a library. That did not work with the bbn nmea sensor project. I looked at the yaml and all of the libraries have been included so I’m baffled about why it will not compile in the IDE.

**Comment by tgsail1** [comment]
My starting point was bbn_atoms3r_mahoney.ino. That seems to mostly work after applying calibration. I did manage to find a forked version of nmea2000_esp32 which supports the esp32-s3. It was kindly provided by t. lapalainnen last year. The entire nmea2k framework is based upon homberger’s work.  I’ve also made an engine hour meter using the same framework. Thanks for your work and your help

T

**Reply by mgrouch** [reply]
In waves QMEKF should give better results. Might need more tuning though

**Comment by tgsail1** [comment]
Just tried this and it does not calibrate very well. I think the magnetic calibration needs to be longer.
T

**Reply by mgrouch** [reply]
How do you measure quality of calibration? How do you perform compass calibration? You need to rotate the device 360 in all possible 3D directions for 45 sections or so.

**Reply by tgsail1** [reply]
I measure the quality two ways: on a flat surface rotate through 90 deg steps to see whether the response is linear (that +-90 and 180 rotations have corresponding differential 90 deg readings within a few degrees. I also compare it to an iphone compass which has performed well. I've tried to calibrate it several ways; Rotation of 360 deg on all three axes on a flat surface, or rotating it in my hands in the air. I've also tried a hybrid of the two. This actually produced the best results. I might make a fixture and try it that way. What might make the calibration simpler is to allow recalibration of the magnetometer alone, without having to go through accel and gyro. Thanks for all the work. This has been a fun project.

**Reply by mgrouch** [reply]
Here is how to build IMU calibration table DIY from PVC. 

https://thecavepearlproject.org/2015/05/22/calibrating-any-compass-or-accelerometer-for-arduino/


For compass calibration step you need to rotate continuously without placing it. 

For accelerometer calibration you need exact poses on all edges. Make sure they are exact 90 degrees increments or gravity will be misaligned and you will see something that you saw with incomplete rotation estimates. 

Are you sure iPhone compass is good?

**Reply by tgsail1** [reply]
Put this together this afternoon and it didn't perform any better

https://github.com/user-attachments/assets/2bb7ab82-9340-4d2a-9187-f897317f7abe

**Comment by mgrouch** [comment]
That doesn't look like it would give enough points on the sphere. You need
to rotate as much as possible in all 3D sphere

**Comment by tgsail1** [comment]
Ok. Finally got a good calibration. The key was to disconnect the USB and use a very long, thin wire on the grove connector, being very careful to keep the wire away from the unit. This gave me access to more angles to try. 3 suggested code improvements: 1. allow more than 45 sec for the mag cal. 2. allow the mag cal to be redone without changing gyro and accel. 3. provide some graphical or numerical feedback to communicate the quality of the mag cal ( simply a metric of angular coverage would help). Thanks again. I’ll try to modify the code to put out heading and imu pgn data.
T
